XWIKI-13786-2: Accessibility problem with rights management's checkboxes states

Sereza7 opened the pull request:
Jira: https://jira.xwiki.org/browse/XWIKI-13786

Continuation of: https://github.com/xwiki/xwiki-platform/pull/542
This old PR was small and the sole file it concerned moved since then, creating a merge issue.

I reused the lines from the above PR to solve the initial issue and added more accessibility support (mentioned by the issue reporter on jira) for the Rights table: keyboard navigation and interactivity.

vmassol commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/js/xwiki/usersandgroups/usersandgroups.js:
```diff
@@ -55,6 +55,13 @@ window.MSCheckbox = Class.create({
+    this.alts = [
+      '$escapetool.javascript($services.localization.render("none"))',
+      '$escapetool.javascript($services.localization.render("allow_1"))',
+      '$escapetool.javascript($services.localization.render("allow_0"))'
```
the keys are missing some namespacing. See https://dev.xwiki.org/xwiki/bin/view/Community/DevelopmentPractices#HTranslationBestPractices

Sereza7 replied:
I decided to use the 'platform.core.rightsManagement.' prefix/namespace. 
I'm not sure I understood how I'm supposed to extend the translations.
Of what I understood, since these string are used in the core project, all I need to do is add them directly in the ApplicationResources.properties file. There is nothing to do with L10N.

`platform.core.rightsManagement.allowed=Allowed
platform.core.rightsManagement.denied=Denied
platform.core.rightsManagement.inherit=Inherit`

michitux replied:
You don't need to do anything for Weblate, that's automatic.

vmassol commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/js/xwiki/usersandgroups/usersandgroups.js:
```diff
@@ -89,6 +99,8 @@ window.MSCheckbox = Class.create({
       delete this.table.fetchedRows[this.idx];
     }
     this.draw(this.state);
+    /*After changing the image through an interaction, put focus back on it.*/
```
This is explaining what it does but not why. In comments, it's always better to explain why we do something (more useful when another coder reads it or when you read it back a few years after :)).

vmassol commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/js/xwiki/usersandgroups/usersandgroups.js:
```diff
@@ -205,10 +217,21 @@ window.MSCheckbox = Class.create({
   attachEvents: function()
   {
     Event.observe(this.domNode, 'click', this.createClickHandler(this));
+    Event.observe(this.domNode, 'keydown', this.createKeyboardHandler(this));   /*Added for accessibility through keyboard*/
```
This is a good comment (it explains why). I'd place it like this:

```
// Added for accessibility through keyboard
Event.observe(this.domNode, 'keydown', this.createKeyboardHandler(this)); 
```

vmassol commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/js/xwiki/usersandgroups/usersandgroups.js:
```diff
@@ -609,4 +632,4 @@ function setGuestExtendedRights(self)
     }
   }
-}
+}
```
the last new line was removed. Was this intentional?

Sereza7 replied:
My bad, I had trouble figuring out what this was about.
I understood why it happened (text mode edits in Linux) and fixed it.

michitux commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/js/xwiki/usersandgroups/usersandgroups.js:
```diff
@@ -71,6 +78,9 @@ window.MSCheckbox = Class.create({
     //add new image
     var img = document.createElement('img');
     img.src = this.images[state];
+    img.alt = this.domNode.getAttribute('data-title') + " " + this.alts[state];
```
Combining several translation messages in a fixed order is a bad practice and incompatible with some languages. If you need to combine messages, one of them should have a placeholder for the other such that the translator can determine its position. See also https://dev.xwiki.org/xwiki/bin/view/Community/L10N/Conventions/#HRulesfortranslatingcompositeandlongmessages

michitux commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/js/xwiki/usersandgroups/usersandgroups.js:
```diff
@@ -71,6 +78,9 @@ window.MSCheckbox = Class.create({
     var img = document.createElement('img');
     img.src = this.images[state];
+    img.alt = this.domNode.getAttribute('data-title') + " " + this.alts[state];
+    img.tabIndex = 0;
```
Isn't it a bad accessibility practice to add focus to elements that don't support it natively? Wouldn't it be better to wrap the image in a button?

Sereza7 replied:
I wrapped the image in a button and moved most of the behavior to this button. :+1:

michitux reviewed (commented):
This whole class that tries to add button-like behavior to an image looks like a huge hack imho. At least, I think an element that supports proper keyboard and mouse interaction natively should be added as a wrapper instead of adding more hacks to it to make it somehow accessible. Possibly, it might also make sense to just deprecate this whole class and replace it by something more modern.

michitux commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/js/xwiki/usersandgroups/usersandgroups.js:
```diff
@@ -55,28 +64,51 @@ window.MSCheckbox = Class.create({
+        l10n['denied']
+      ];
+      img.alt = alts[this.state];
+      this.button.title = alts[state];
```
I wonder if it is enough to indicate the state. On the one hand, it could be annoying for screen reader users to have more text, on the other hand, how do they know what clicking on the button does? Do you know about best practices for this situation?

michitux replied:
I think the cleanest thing would be to make the items of the table select inputs as this really models what this is about. The problem is that styling of select elements is extremely limited in HTML and we would be forced to put plain text values in there which would be a significant change and not so nice for the design I fear.

What you suggest sounds good, I think putting the current state first should allow screen reader users to quickly switch to the next item in the table after they heard the first word, but maybe it could be good to confirm this with an actual screen reader user (if this is what they would prefer or if there should, e.g., be some instructions that describe the buttons generically at the start of the table and just the current state on the actual buttons).

michitux replied:
Btw., I think the respective reference documentation for accessibility of this kind of button is https://www.w3.org/WAI/tutorials/images/functional/#example-5-image-used-in-a-button which says that the `alt` attribute should be used to "convey the purpose of the button".

michitux commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/resources/ApplicationResources.properties:
```diff
@@ -2109,6 +2109,9 @@ platform.core.rightsManagement.editRightsForSpace=Editing access rights for spac
 platform.core.rightsManagement.saveComment={0} {1} right for {2}
+platform.core.rightsManagement.allowed=Allowed
+platform.core.rightsManagement.denied=Denied
+platform.core.rightsManagement.inherit=Inherit
```
This is more complex than inherit. For example, when you give one group view right on a document, the view right is denied for all other groups, the value of the parent has no influence here. It is just not set I would say.

manuelleduc replied:
wdyt of `undefined` for the "not set values"?

manuelleduc commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/js/xwiki/usersandgroups/usersandgroups.css:
```diff
@@ -50,6 +50,10 @@
 #usersandgroupstable.rights .rights {
   width: 15% !important;
 }
+#usersandgroupstable button.rights-edit {
```
If this is done for accessibility reason. Can you add a comment explaining what's the impact of this block on accessibility?

manuelleduc reviewed (commented):
General remark: it is often useful for reviewers to have a comment explaining what changes with the PR (and if the change is visual, having before/after screenshot helps a lot - or video recordings)

Sereza7 commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/js/xwiki/usersandgroups/usersandgroups.js:
```diff
@@ -55,28 +64,51 @@ window.MSCheckbox = Class.create({
+    
+    this.label = document.createElement('span');
+    this.label.hidden = true;
+    this.label.appendChild(document.createTextNode(this.labels[this.state]));
```
The label element was present in the structure, however it was never used.
I changed the way the table was generated, so in the case one wanted to use it someday, I had to also change the way the label element was instantiated (and updated).

We should be able to remove it without any changes, it remains hidden.

michitux replied:
I would actually be inclined to just remove these label elements as with the alt text and title they are quite redundant though it could be considered that they are part of the API of XWiki which shouldn't be changed. Any other opinions on this?

mflorea replied:
I don't think this label was ever used, so I'm in favor or removing it.

Sereza7 commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/js/xwiki/usersandgroups/usersandgroups.js:
```diff
@@ -93,13 +94,13 @@ window.MSCheckbox = Class.create({
         l10n['denied']
       ];
-      img.alt = alts[this.state];
+      img.alt = alts[this.state] + " . " + l10n.get('changeStateTo', alts[this.nextState()]);
```
I meant for the dot to just be a separator between two different semantic units juxtaposed (current state & purpose of the button).
Okay, got it. I changed the translation in the three states to just contain one string that also defines the next state (since this should not change, I guess it can be hard written in the translations. **Should I add a commentary for these entries in the application resource or is that good as is?**
There's a couple needs/preferences for the translation that might get lost otherwise, but I navigated through the resource bundle and could hardly see any comment except those for delimiting sections.
Something like:
### Needs to define current state and then purpose of the button. Should start with the current state name for easier navigation with screen readers.

michitux replied:
We rarely add comments, yes, but it would be a good idea to change that. Comments above the translation string are displayed in Weblate afaik so this should work.

Sereza7 commented:
PR Changes:
Fixed alt text on image and title of button. The new alt text is not composed. Most of the complexity is shifted to the L10N entry itself. The new alt text is composed of the button status followed by its purpose (switching status).
The third status has been changed to 'undefined' instead of 'inherit'. 
Fixed a lot of issues with L10N usage in the .js file.

This screenshot displays the result on a local instance of WCAG2.1AA validation using the Wave extension for Firefox.
![WaveAnalysis](https://user-images.githubusercontent.com/28761965/215052799-ab1cd3bb-aa8f-4907-8ffc-8ee2003f7e6c.png)
We can see that this validation does not display any violation in the rights edit table, and that the button titles are properly set.



For reference, here is the tabbing order through the table (displayed with Firefox Inspector accessibility tool):
![tabbingOrderRights](https://user-images.githubusercontent.com/28761965/215052796-af0f8c8f-d400-4ebb-9e61-61b33e05fb21.png)

michitux commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/js/xwiki/usersandgroups/usersandgroups.js:
```diff
@@ -55,34 +64,62 @@ window.MSCheckbox = Class.create({
+    
+    this.button = document.createElement("button");
+    this.button.className = "rights-edit";
+    this.button.onclick = this.createClickHandler(this);
```
This looks like a quite old way to add event listener, wouldn't it be better to use [`addEventListener`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener)? I don't know what the best practices are, but may feeling is that the current way is not a best practice.

michitux commented on xwiki-platform-core/xwiki-platform-oldcore/src/main/resources/ApplicationResources.properties:
```diff
@@ -2109,6 +2109,9 @@ platform.core.rightsManagement.editRightsForSpace=Editing access rights for spac
 platform.core.rightsManagement.saveComment={0} {1} right for {2}
+platform.core.rightsManagement.allowed=Allowed is the current state. Switch to denied.
+platform.core.rightsManagement.denied=Denied is the current state. Switch to undefined.
+platform.core.rightsManagement.undefined=Undefined is the current state. Switch to allowed.
```
What do you think about something like "Allowed. Click to deny."/"Denied. Click to switch to undefined."/"Undefined. Click to allow." I'm not sure about the exact wording, either, but I think it could be interesting to have a shorter text when possible.

michitux reviewed (commented):
I have still some small comments but overall I think this looks good now.

michitux commented on xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/js/xwiki/usersandgroups/usersandgroups.js:
```diff
@@ -55,34 +64,62 @@ window.MSCheckbox = Class.create({
+    
+    this.button = document.createElement("button");
+    this.button.className = "rights-edit";
+    Event.observe(this.button, 'click', this.createClickHandler(this));
```
Is there any reason for using Prototype here? We're moving away from Prototype so I don't think it is a good idea to expand it's usage unless it is strictly required (see also, e.g., https://jira.xwiki.org/browse/XWIKI-17645).

Sereza7 replied:
Thanks! I made the changes too quickly, I reverted it back to the way they added the listener before this PR, and didn't realize it was still different from the way you linked [above](https://github.com/xwiki/xwiki-platform/pull/2051#discussion_r1088768833).

Sereza7 commented:
**PR Changes:**

Updated L10N key en values to shorter strings. 
Added resource comments to emphasize the importance of keeping the two elements in the key.
Fixed the listener instantiation.

___
Might need more changes to resolve [this conversation](https://github.com/xwiki/xwiki-platform/pull/2051#discussion_r1086906531).
The problem is that the former structure of the table cell had the possibility to keep and update a label. However, this label was not used at all. The question at hand is whether or not to remove this (unused) label.

Sereza7 commented:
**PR Changes:**

- Removed the (unused) label from the rights-edit. See https://github.com/xwiki/xwiki-platform/pull/2051#discussion_r1086354379
- Fixed a comment in the css
- Fixed a bad entry in the applicationResources

**View:**

No changes since https://github.com/xwiki/xwiki-platform/pull/2051#issuecomment-1406249803
